activation/activation.go

```diff
@@ -195,35 +240,64 @@ func (b *Builder) loop(ctx context.Context) {
 	if err != nil {
 		log.Info("challenge not loaded: %s", err)
 	}
-	if err := b.waitOrStop(b.initDone); err != nil {
+	if err := b.waitOrStop(ctx, b.initDone); err != nil {
 		return
 	}
 	// ensure layer 1 has arrived
-	if err := b.waitOrStop(b.layerClock.AwaitLayer(1)); err != nil {
+	if err := b.waitOrStop(ctx, b.layerClock.AwaitLayer(1)); err != nil {
 		return
 	}
+	b.run(ctx)
+}
+
+func (b *Builder) run(ctx context.Context) {
+	var poetRetryTimer *time.Timer
+	defer func() {
+		if poetRetryTimer != nil {
+			poetRetryTimer.Stop()
+		}
+	}()
+	defer b.log.Info("atx builder is stopped")
 	for {
-		select {
-		case <-b.stop:
-			return
-		default:
+		client := atomic.LoadPointer(&b.pendingPoetClient)
```

Comment:
consider encapsulating poet `UpdatePoETServer` within poetClient. As I see it there is no need for this functionality to be part AtxBuilder as it does not use it internally, we can export this functionality to API from poet client wrapper.

Reply:
this is in AtxBuilder because it controls the concurrency of the execution. We can't update poet just at any time, because It can be used several times during build process, and we know that it is not used only before execution is started.

for example here https://github.com/spacemeshos/go-spacemesh/blob/develop/activation/nipst.go#L162-L174

we first fetch PoetServiceID, then Submit and after that, we are waiting for a proof from a PoetServiceID that we fetched. If it was updated after PoetServiceID is fetched we will just timeout later. Besides that, if we are going to rotate between multiple clients this will be the only safe place to do that as well.

I considered doing some locking in NiPSTBuilder, but it in results in error-prone code. Much safer to assume that we always interact with the same client until an error has occurred.